lib/opal/nodes/case.rb

```diff
@@ -8,34 +8,39 @@ class CaseNode < Base
       children :condition
 
       def compile
-        handled_else = false
-
         compiler.in_case do
-          if condition
-            case_stmt[:cond] = true
-            add_local '$case'
+          compile_code
 
-            push "$case = ", expr(condition), ";"
-          end
+          wrap '(function() {', '})()' if needs_closure?
+        end
+      end
 
-          case_parts.each_with_index do |wen, idx|
-            if wen and wen.type == :when
-              compiler.returns(wen) if needs_closure?
-              push "else " unless idx == 0
-              push stmt(wen)
-            elsif wen # s(:else)
-              handled_else = true
-              wen = compiler.returns(wen) if needs_closure?
-              push "else {", stmt(wen), "}"
-            end
-          end
+      def compile_code
+        handled_else = false
 
-          # if we are having a closure, we must return a usable value
-          if needs_closure? and !handled_else
-            push "else { return nil }"
+        if condition
+          case_stmt[:cond] = true
+          add_local '$case'
+
+          push "$case = ", expr(condition), ";"
+          empty_line
+        end
+
+        case_parts.each_with_index do |wen, idx|
+          if wen and wen.type == :when
+            wen = compiler.returns(wen) if needs_closure?
+            push "else " unless idx == 0
+            push stmt(wen)
+          elsif wen # s(:else)
```

Comment:
I'm not sure I like this.

I think it would be better to check `if wen` and then do a `case`.